File: ontrack-extension-jira/src/main/java/net/nemerosa/ontrack/extension/jira/tx/JIRASessionFactoryImpl.java
package net.nemerosa.ontrack.extension.jira.tx;

import net.nemerosa.ontrack.client.JsonClient;
import net.nemerosa.ontrack.extension.jira.JIRAConfiguration;
import net.nemerosa.ontrack.extension.jira.client.JIRAClient;
import net.nemerosa.ontrack.extension.jira.client.JIRAClientImpl;
import net.nemerosa.ontrack.extension.support.client.ClientConnection;
import net.nemerosa.ontrack.extension.support.client.ClientFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

@Component
public class JIRASessionFactoryImpl implements JIRASessionFactory {

    private final ClientFactory clientFactory;

    @Autowired
    public JIRASessionFactoryImpl(ClientFactory clientFactory) {
        this.clientFactory = clientFactory;
    }

    @Override
    public JIRASession create(JIRAConfiguration configuration) {
        // Creates a HTTP JSON client
        JsonClient jsonClient = clientFactory.getJsonClient(
                new ClientConnection(
                        configuration.getUrl(),
                        configuration.getUser(),
                        configuration.getPassword()
                )
        );
        // Creates the client
        JIRAClient client = new JIRAClientImpl(jsonClient);
        // Creates the session
        return new JIRASession() {

            @Override
            public JIRAClient getClient() {
                return client;
            }

            @Override
            public void close() {
                client.close();
            }
        };
    }

}

File: ontrack-extension-jira/src/main/java/net/nemerosa/ontrack/extension/jira/tx/JIRASessionFactoryImpl.kt
package net.nemerosa.ontrack.extension.jira.tx

import net.nemerosa.ontrack.common.RunProfile
import net.nemerosa.ontrack.extension.jira.JIRAConfiguration
import net.nemerosa.ontrack.extension.jira.client.JIRAClient
import net.nemerosa.ontrack.extension.jira.client.JIRAClientImpl
import net.nemerosa.ontrack.extension.support.client.RestTemplateBasicAuthentication
import net.nemerosa.ontrack.extension.support.client.RestTemplateProvider
import org.springframework.context.annotation.Profile
import org.springframework.stereotype.Component

@Component
@Profile("!${RunProfile.DEV}")
class JIRASessionFactoryImpl(
    private val restTemplateProvider: RestTemplateProvider,
) : JIRASessionFactory {

    override fun create(configuration: JIRAConfiguration): JIRASession {

        // Creates an HTTP client
        val template = restTemplateProvider.createRestTemplate(
            rootUri = configuration.url,
            basicAuthentication = RestTemplateBasicAuthentication(
                username = configuration.user ?: "",
                password = configuration.password ?: "",
            )
        )

        // Creates the client
        val client: JIRAClient = JIRAClientImpl(template)

        // Creates the session
        return object : JIRASession {

            override fun getClient(): JIRAClient {
                return client
            }

            override fun close() {
                client.close()
            }
        }
    }
}
